### DISCUSSION TITLE
Cannot push to remote Git repo

### DISCUSSION THREAD

**Discussion by maxxer** [WeblateOrg/weblate#6651] [Q&A]
Hi.
I've set up my weblate component to pull from a Gitlab repo's master branch and push to a `weblate` branch. Every time I make a translation change and try to push to Gitlab, Weblate locks the project with this error:

```
To ssh://my.gitlab.host:22/repo.git
 ! [rejected]        master -> weblate (non-fast-forward)
error: failed to push some refs to 'ssh://my.gitlab.host:22/repo.git'
hint: Updates were rejected because a pushed branch tip is behind its remote
hint: counterpart. Check out this branch and integrate the remote changes
hint: (e.g. 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

No matter what. If I delete the `weblate` branch from the repo it successfully pushes once, but the second time it throws this error. 

I tried to figure out what's going on but wasn't able to sort it out.

Any suggestion?
Thanks

**Comment by nijel** [comment]
If you use rebase strategy in Weblate, you will need a force push here - because the Weblate history is rewritten on each repository update.

**Reply by maxxer** [reply]
So basically the `weblate` branch always has only one new commit every time a translation change is made?

**Reply by nijel** [reply]
Please read https://docs.weblate.org/en/latest/admin/continuous.html#lazy-commits for more details.

**Reply by maxxer** [reply]
I've checked this and the documentation about the Git repository, but I couldn't find what's wrong.

I've configured the same Git repo for pull and push, but a different branch for push. Every time there's a new update on translation Weblate complains it cannot push to the remote `weblate` branch because *because a pushed branch tip is behind its remote
hint: counterpart*. It doesn't seem to me a matter of lazy commits...

**Reply by nijel** [reply]
In case you are using rebase [Merge style](https://docs.weblate.org/en/latest/admin/projects.html#merge-style), you might want to use [force push](https://docs.weblate.org/en/latest/vcs.html#git-with-force-push) together with dedicated push branch.

I've added this hint to the alert in f9bcc58d505136233571aecd07c35c1911551f1e.

**Reply by maxxer** [reply]
I tried both _merge_ and _rebase_ and they both seem to behave the same. I'll go with *force push*, I didn't search well enough for it

**Reply by nijel** [reply]
Changing to merge will not change the history - you will still have the rebased commits while the pushed branch has them not rebased. So while it will address it for the future, it can't fix it in case the problem is already there.

**Comment by sebpeterson** [comment]
I have the same issue. How can I issue the "git pull" from within the docker container ? Where is the git structure within the container ?

**Reply by nijel** [reply]
See https://docs.weblate.org/en/latest/admin/install/docker.html#docker-container-volumes and https://docs.weblate.org/en/latest/admin/config.html#std-setting-DATA_DIR

**Reply by bmaehr** [reply]
Would it be possible to add a symbolic link
``` 
ln -s /app/data/ssh /home/weblate/.ssh
``` 
which would it make much easier to do something with git in the container?

**Reply by nijel** [reply]
This is a bit more tricky as Weblate uses custom HOME as well, so this would change only SSH related settings.

**Reply by bmaehr** [reply]
Do you have a better idea? (In fact I would prefer not to have to connect to the docker to do some manual cleanup and would like to use some button in the UI instead. I tried all of them, none of them did solve it)

**Reply by nijel** [reply]
You should not need to connect to the Docker container. What are you trying to solve?

**Reply by bmaehr** [reply]
The issue of this thread. (I now already solved it by creating a directory with a link to .ssh at /tmp and modifying the HOME varibale to point there).
